### ISSUE TITLE
Improvement: Make all nodes callable

### ISSUE THREAD

**Issue by ThomasMahon** [DynamoDS/Dynamo#6510]
## Dynamo version

0.9.2

## Operating system

Win 10

##Improvement Suggestion
Make nodes callable (i.e force users to give each node placed on the graph a unique name)

**Comment by ksobon** [comment]
nodes in Dynamo are already callable by virtue of the methods running behind them being available via Design Script in Code Blocks. Now, not all of them are exposed yet in Design Script but that's a work in Progress. On the other hand your suggestion to make users enter unique name for all new nodes, is just slightly far fetched and almost impossible to orchestrate. It would also be a massive hit on usability. -1

**Comment by ThomasMahon** [comment]
Hi Konrad

I mean calling nodes already on the graph which is why each would need a unique name. If I place a slider for example, I want to call it elsewhere in my script without having to redefine it. Sure, I can use an input and wire it up, but what if I've used a number of nodes to generate a result and want whichever node has created that result to be the one I call. I don't want to recreate all of the dependent nodes again. I acknowledge that this suggestion would break every single Dynamo file out there but I don't agree that this would impact usability; all Dynamo needs to do is name nodes generically as they are placed on the graph. Say I place a few point nodes, they are named sequentially without me needing to do anything: point01, point02, etc. 

Implementing this functionality would open up so many benefits: 
- Enforce naming conventions
- Find...and even replace: Take my slider, what if I want to replace it but it's connected downstream to 10 other nodes. It's a pain trying to rewire particularly on large definitions. Imagine the power of a replace command
- Increased efficiency particularly in DesignScript
- Fully integratable debugging 
- Enable more concepts for object association. Ultimately it will solve some of the headaches that arise with visual programming 

No chance before version 1? ;)

**Comment by josols** [comment]
I agree with Konrad. I know this is not the forum, and although it is a little workaroundish, you could achieve less wiring by making callable definitions for your variables.
![image](https://cloud.githubusercontent.com/assets/7080648/14777889/397008f4-0ad0-11e6-869d-2368c26e0335.png)

**Comment by ThomasMahon** [comment]
Absolutely. However the underlying wish is to be able to call nodes by name once placed on the graph. Using definitions gets no closer to this, especially when scrutinised against the list of benefits I mentioned - find replace being probably the most significant gain.

**Comment by josols** [comment]
As per Twitter, I would like to see something the like of the ProruBIM package, though it only accepts the standard inputs available in Dynamo. Something the like of below image:
![image](https://cloud.githubusercontent.com/assets/7080648/14885756/8839ef84-0d4d-11e6-8144-8ae7769d0b24.png)

**Comment by ThomasMahon** [comment]
This would definitely be a step in the right direction. In a professional team-based environment (i.e. architectural practices) where there are multiple authors working on large definitions, enforcing naming conventions is the only sensible way to manage the process and collaborate - I'm a big advocate as a result, and therefore see unique names - which encourage good naming standards - as a great benefit.

**Comment by ke-yu** [comment]
@ThomasMahon , I think the callable node here is code block node, right? We've considered the possibility of compiling a code block node to an function, probably anonymous so that if a code block node hasn't get all inputs, it could be used as partially applied function and be connected to high order function like `List.Map`. If giving it a name, then it is callable. 

But the major issue is usability, unless we have the other kind of node which doesn't have any output and behaves as a function. If we go this direction, then probably a general function definition node makes more sense.

**Comment by ThomasMahon** [comment]
Yes absolutely, your suggestion sounds interesting. However, I was also thinking beyond this - along the lines of having every node uniquely named to enable the implementation of find/replace functionality which would be very useful, particularly on large graphs. Also, from a DesignScript standpoint, if the node is uniquely named I can call it explicitly within the body of the code rather than writing an input argument (essentially, the nodes become global variables). I appreciate this wouldn't necessarily be regarded as good coding practice, but there are some cases where this is sensible (simplifying packaged nodes for team-based working for example). Also, it would enable a 'wireless' link to exist which would help de-clutter busy graphs. Look at this script I'm currently developing for example - 'spaghetti junction' helps no one:

![capture](https://cloud.githubusercontent.com/assets/18601851/14941540/6393a8c2-0f97-11e6-87a2-c54c9e151881.png)

**Comment by dimven** [comment]
Each node placed on the canvas already has a unique id. That's how the graph tells identical nodes apart... It'd be much better to expose the GUID through some back-end referencing mechanism than to trouble the user of adding a unique name on placement.

**Comment by ThomasMahon** [comment]
That would certainly be good practice and custom nodes help, however when confronted by a large graph, for arguments sake with say multiple List.Map nodes (I have a graph that does incidentally), without renaming them to something more meaningful like: ptPrimaryStructuralOrigins, ptLandingOrigins, etc the graph quickly becomes almost meaningless and unusable, so renaming nodes is not only best-practice, but really inevitable to avoid making life hell. Notes and groups simply enhance this practice. With that in mind, I don't see renaming nodes as a burden to users. My suggestion is for Dynamo to name each new node sequentially as placed on the graph to avoid burdening the user (and go largely unnoticed). As I mentioned, naming nodes with meaningful names is a great benefit which keeps a project running smoothly (...obviously only if good naming conventions are in place!).

**Comment by dimven** [comment]
To further clarify what i meant in my comment above : _"...might impede of the readability of the graph..."_ - what I meant to say was that it might affect the overall comprehension, or flow,  of the graph. I.e. what's a good way to visually identify where the wireless connections are enforced? Do you also need to differentiate multiple outbound connectors? If I have multiple wireless connectors going out, and need to change only one of them how do I do it?

Basically, you'll need to redefine the entire action of connecting nodes and figure out how to present all of the information in a simple, easy to understand and robust manner.

**Comment by ThomasMahon** [comment]
I did understand you and you're absolutely right - could be easy to fix: click the receiving node and a 'ghost' wire displays showing the dependency back to the source. Grasshopper is a good precedent for this behaviour when hiding wires for example. 

Wireless nodes are side issue though - I want to remain on topic: 1. callable nodes (that are present on the graph) would be highly beneficial for all the reasons I've put forward. 2. As a result all nodes would require unique names.

**Comment by Curty102** [comment]
Delayed additions to this topic--

Assuming these additions don't slow performance appreciably:

-Agree with Thomas all nodes should be callable in DS/code blocks.
-Agree autonaming should be default to keep software usable for beginners. Variable/node names can be augmented down the pipe as needed to have more meaning.  This is basically an expansion of the autonaming in the node-to-code.
-Alphanumeric rules for variable naming would need to be resolved with node name restrictions, either directly (one set of rules for both --my preference) or by proxy (node names are converted to be DS friendly --not my fav, creates ambiguity.)
-also to maintain usability, this default name/custom name should be in a colapsable menu above the generic node name (ie Point.ByCoordinates) to keep visual data to a minimum by default. Top edge of node could have a different color if name has been changed from default, and could be click-expanded to have all name data at the top visually.
-this same functionality should be implemented for custom nodes upon creation as well.
-Agree with Thomas, Node Replace is brilliant. DS should take it's cue from this and allow an option, within textual script editing, for variable renaming to either be per Instance (thus creating a new variable-this is the default behavior and why wires disconnect from code blocks when variables are renamed) or Global (perhaps via right-click) (this would change the name of that variable in all instances wherever it is referenced in the script. 
-And as for wireless or hidden node connections -   Take cue from Grasshopper, wires changed to be hidden (via right-click) show up as screened upon node selection.  Improvement to THAT: upon a node selection, highlight all directly dependant Nodes in graph AND variables in DS.  This could be a UI option to turn on/off, in varying degrees, to strike a balance between performance and troubleshooting abilities/preference. 
-Agree with Jostein, Prorubim wireless variable output nodes are very useful. This would inherantly be achieved by implimentation of all above.

**Comment by ke-yu** [comment]
@Racel
